blib/lib/Sys/Statistics/linux.man
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "Linux 3"
.TH Linux 3 "2006-12-09" "perl v5.8.4" "User Contributed Perl Documentation"
.SH "NAME"
Sys::Statistics::Linux \- Collect linux system statistics.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&   use Sys::Statistics::Linux;
.Ve
.PP
.Vb 1
\&   my $lxs = new Sys::Statistics::Linux;
.Ve
.PP
.Vb 14
\&   $lxs\->set(
\&      SysInfo   => 1,
\&      CpuStats  => 1,
\&      ProcStats => 1,
\&      MemStats  => 1,
\&      PgSwStats => 1,
\&      NetStats  => 1,
\&      SockStats => 1,
\&      DiskStats => 1,
\&      DiskUsage => 1,
\&      LoadAVG   => 1,
\&      FileStats => 1,
\&      Processes => 1,
\&   );
.Ve
.PP
.Vb 1
\&   sleep 1;
.Ve
.PP
.Vb 1
\&   my $stat = $lxs\->get;
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This module is the main package from the distribution Sys::Statistics::Linux and collects
different linux system informations like processor workload, memory usage, network and
disk statisitcs and other system informations. Refer to the documentation of the distribution
modules to get more informations about all possible statistics and system informations.
.SH "TECHNICAL NOTE"
.IX Header "TECHNICAL NOTE"
This distribution collects statistics by the virtual \fI/proc\fR filesystem (procfs) and is developed
on default vanilla kernels. It is tested on x86 hardware with the distributions SuSE (SuSE on s390
and s390x architecture as well), Red Hat, Debian, Asianux, Slackware and Mandrake on kernel versions
2.4 and 2.6 and should run on all linux kernels with a default vanilla kernel as well. It is possible
that this module doesn't run on all distributions if the procfs is too much modified.
.PP
Further it is necessary to run it as a user with the authorization to read the \fI/proc\fR filesystem.
.SH "DELTAS"
.IX Header "DELTAS"
The options \f(CW\*(C`CpuStats\*(C'\fR, \f(CW\*(C`ProcStats\*(C'\fR, \f(CW\*(C`PgSwStats\*(C'\fR, \f(CW\*(C`NetStats\*(C'\fR, \f(CW\*(C`DiskStats\*(C'\fR and \f(CW\*(C`Processes\*(C'\fR
are deltas, for this reason it's necessary to initialize the statistics first, before the data
be generated with \f(CW\*(C`get()\*(C'\fR. The statistics can be initialized with the methods \f(CW\*(C`new()\*(C'\fR, \f(CW\*(C`set()\*(C'\fR
and \f(CW\*(C`init()\*(C'\fR. Each option that is set to \s-1TRUE\s0 (1) will be initialized by the call of \f(CW\*(C`new()\*(C'\fR
or \f(CW\*(C`set()\*(C'\fR. The call of \f(CW\*(C`init()\*(C'\fR reinitialize all statistics that are set to 1. By the call
of \f(CW\*(C`get()\*(C'\fR the initial statistics will be updated automatically. Please refer the \s-1METHOD\s0 section
to get more information about the calls of \f(CW\*(C`new()\*(C'\fR, \f(CW\*(C`set()\*(C'\fR and \f(CW\*(C`get()\*(C'\fR.
.PP
Another exigence is that you need to sleep for while \- at least for one second \- before you
call \f(CW\*(C`get()\*(C'\fR if you want to get useful statistics. The options \f(CW\*(C`SysInfo\*(C'\fR, \f(CW\*(C`MemStats\*(C'\fR,
\&\f(CW\*(C`SockStats\*(C'\fR, \f(CW\*(C`DiskUsage\*(C'\fR, \f(CW\*(C`LoadAVG\*(C'\fR and \f(CW\*(C`FileStats\*(C'\fR are no deltas. If you need only one
of this informations you don't need to sleep before the call of \f(CW\*(C`get()\*(C'\fR.
.PP
The \f(CW\*(C`get()\*(C'\fR function collects all requested informations and returns a hash reference with the
statistics. The inital statistics will be updated. You can turn on and off options with \f(CW\*(C`set()\*(C'\fR.
.SH "OPTIONS"
.IX Header "OPTIONS"
All options are identical with the package names of the distribution. To activate the gathering
of statistics you have to set the options by the call of \f(CW\*(C`new()\*(C'\fR or \f(CW\*(C`set()\*(C'\fR. In addition you
can deactivate \- respectively delete \- statistics with \f(CW\*(C`set()\*(C'\fR and re-init all statistics with
\&\f(CW\*(C`init()\*(C'\fR.
.PP
The options must be set with a \s-1BOOLEAN\s0 value (1|0).
.PP
.Vb 2
\&   1 \- activate (initialize)
\&   0 \- deactivate (delete)
.Ve
.PP
To get more informations about each option refer the different modules of the distribution.
.PP
.Vb 12
\&   SysInfo     \-  Collect system informations             with L<Sys::Statistics::Linux::SysInfo>.
\&   CpuStats    \-  Collect cpu statistics                  with L<Sys::Statistics::Linux::CpuStats>.
\&   ProcStats   \-  Collect process statistics              with L<Sys::Statistics::Linux::ProcStats>.
\&   MemStats    \-  Collect memory statistics               with L<Sys::Statistics::Linux::MemStats>.

ChangeLog
0.09_14 Released at 2007-04-13.
        Code fixes.
0.09_13 Released at 2007-04-13.
        Did some fixes in DiskStats.pm, because the error message
          "... no diskstats found ..." is ever issued.
        Did some fixes in fproc() and search() of Linux.pm.
0.09_12 Released at 2007-04-12.
        Changed search() and fproc(). Now statistics can be set
          as second argument. The first argument have to be the
          search filter. Both arguments have to be a hash refrence.
0.09_11 Released at 2007-04-12.
        To much files uploaded. :-)
0.09_10 Released at 2007-04-12.
        Fixed the "Can't call method _struct" bug in Linux.pm line 419.
        "new" from ProcStats is float now. Did some fixes in ProcStats,
          NetStats, PgSwStats and DiskStats... wrong averages/s calculated.
0.09_09 Released at 2007-04-12.
        Delete key "pid" from Processes() because it exists double
          times: $stats->{Processes}->{$pid}->{pid} = $pid
        Now all values of PgSwStats, NetStats, DiskStats and ProcStats
          (here only the key "new") are values per second. I break
          this down because it's very confusing if the statistics are
          just add together. Now there are all rxbyt/s or pgpgin/s or new/s.
        I kicked IO::File and the _load() Method from some modules
          because they wasn't really necessary.
0.09_08 Released at 2007-03-05.
        Add the method fproc() to search for processes.
        Add t/14fproc.t and changed other test file names.
        Improve the search() method and add a bit documentation.
0.09_06 Released at 2007-02-25.
        Improve the search() method. Now it's possible to search()
          for special devices and more.
0.09_05 Released at 2007-02-23.
        Added realfreeper to MemStats.
0.09_04 Released at 2007-02-23.
        The next release today. :)
        Added key realfree to MemStats (free + buffers + cached).
0.09_03 Released at 2007-02-23.
        Added key ttime to Processes (stime + utime).
        Fixed NetStats.t and Processes.t - missed ttbyt and ttime.
0.09_02 Released at 2007-02-23.
        Fixed search(). Now it returns only that statistics that
          matched the search string.
0.09_01 Released at 2007-02-19.
        A lot of changes in Linux.pm.
        Sub modules will only be loaded if they are needed now.
        Added a search() method to search for values in statistics.
        Now there are different options (-1, 0, 1, 2) available
          for the statistics.
        Added key ttbyt to NetStats (total bytes rx + tx).
0.08    Released at 2007-02-13.
        Added key writeback to MemStats and actualize the
          documentation. The new keys dirty, slab, mapped
          and writeback are only available by kernels from
          2.6.
0.07    Released at 2007-02-12.
        Added keys dirty, slab and mapped to MemStats.
0.06    Released at 2007-01-22.
        Bugfix in ProcStats.pm. "count" statistics fixed.
        Added t/Pod-Coverage.t.
0.05    Released at 2006-12-30.
        Updated t/DiskStats.t.
        Updated POD and README.
0.04    Released at 2006-12-29.
        DiskStats died now if the read access to /proc/partitions
          was successful and if there are no statistics. The reason
          is that it seems to be that the linux kernel (<=2.4) is not
          compiled with CONFIG_BLK_STATS=y. Up to now the message
          "there are no initial statistics defined" was generated
          and this message is not very helpful if init() was called.
0.03    Released at 2006-12-28.
        Updated Sys::Statistics::Linux::CpuStats. There wasn't
          irq and softirq calculated for the uptime for each cpu.
          This values are only available with kernels > 2.4.
          That aren't new keys! Just added for calculation.
          Thanks to Alexey A Makeev for his report!
          In addition I updated the documentation for iowait and
          note that this statistic is only available by kernel
          versions higher than 2.4.

lib/Sys/Statistics/Linux.pm
=head1 NAME

Sys::Statistics::Linux - Front-end module to collect system statistics

=head1 SYNOPSIS

   use Sys::Statistics::Linux;

   my $lxs = new Sys::Statistics::Linux;

   $lxs->set(
      SysInfo   => 1,
      CpuStats  => 1,
      ProcStats => 1,
      MemStats  => 1,
      PgSwStats => 1,
      NetStats  => 1,
      SockStats => 1,
      DiskStats => 1,
      DiskUsage => 1,
      LoadAVG   => 1,
      FileStats => 1,
      Processes => 1,
   );

   sleep 1;

   my $stat = $lxs->get;

=head1 DESCRIPTION

Sys::Statistics::Linux is the front-end module to Sys-Statistics-Linux and collects
different linux system informations like processor workload, memory usage, network and
disk statistics and a lot more. Refer to the documentation of the distribution
modules to get informations about all possible statistics.

=head1 TECHNICAL NOTE

This distribution collects statistics by the virtual F</proc> filesystem (procfs) and is developed
on default vanilla kernels. It is tested on x86 hardware with the distributions SuSE (SuSE on s390
and s390x architecture as well), Red Hat, Debian, Asianux, Slackware and Mandrake on kernel versions
2.4 and 2.6 and should run on all linux kernels with a default vanilla kernel as well.
It is possible that this module doesn't run on all distributions if the procfs is too much modified.

For example the linux kernel 2.4 can compiled with the option "CONFIG_BLK_STATS". It is possible to
activate or deactivate the block statistics for devices with this option. These statistics doesn't
exist in /proc/partitions if this option isn't activated. Since linux kernel 2.5 these statistics are
in /proc/diskstats.

=head1 DELTAS

The options C<CpuStats>, C<ProcStats>, C<PgSwStats>, C<NetStats>, C<DiskStats> and C<Processes>
are deltas, for this reason it's necessary to initialize the statistics first before the data
be generated with C<get()>. These statistics can be initialized with the methods C<new()>, C<set()>
and C<init()>. Each option that is set to TRUE (1) will be initialized by the call of C<new()>
or C<set()>. The call of C<init()> reinitialize all statistics that are set to 1. By the call
of C<get()> the initial statistics will be updated automatically. Please refer the METHOD section
to get more information about the calls of C<new()>, C<set()> and C<get()>.

Another exigence is that you need to sleep for while - at least for one second - before you
call C<get()> if you want to get useful statistics. The options C<SysInfo>, C<MemStats>,
C<SockStats>, C<DiskUsage>, C<LoadAVG> and C<FileStats> are no deltas. If you need only one
of this informations you don't need to sleep before the call of C<get()>.

The C<get()> function collects all requested informations and returns a hash reference with the
statistics. The inital statistics will be updated. You can turn on and off options with C<set()>.

=head1 OPTIONS

All options are identical with the package names of the distribution. To activate the gathering
of statistics you have to set the options by the call of C<new()> or C<set()>. In addition you
can delete, set pause or create new statistics with C<set()> and re-init all statistics with
C<init()>.

The options must be set with on of the following values:

   -1 - set pause on statistics but wouldn't delete them
    0 - delete statistics and destroy the object
    1 - create a new object and init statistics if necessary
    2 - create a new object if not exists but wouldn't init statistics

To get more informations about the statistics refer the different modules of the distribution.

   SysInfo     -  Collect system informations             with Sys::Statistics::Linux::SysInfo.
   CpuStats    -  Collect cpu statistics                  with Sys::Statistics::Linux::CpuStats.
   ProcStats   -  Collect process statistics              with Sys::Statistics::Linux::ProcStats.
   MemStats    -  Collect memory statistics               with Sys::Statistics::Linux::MemStats.
   PgSwStats   -  Collect paging and swapping statistics  with Sys::Statistics::Linux::PgSwStats.
   NetStats    -  Collect net statistics                  with Sys::Statistics::Linux::NetStats.
   SockStats   -  Collect socket statistics               with Sys::Statistics::Linux::SockStats.
   DiskStats   -  Collect disk statistics                 with Sys::Statistics::Linux::DiskStats.
   DiskUsage   -  Collect the disk usage                  with Sys::Statistics::Linux::DiskUsage.
   LoadAVG     -  Collect the load average                with Sys::Statistics::Linux::LoadAVG.
   FileStats   -  Collect inode statistics                with Sys::Statistics::Linux::FileStats.
   Processes   -  Collect process statistics              with Sys::Statistics::Linux::Processes.

=head1 METHODS

=head2 new()

Call C<new()> to create a new Sys::Statistics::Linux object. You can call C<new()> with options.
This options would be hand off to the C<set()> method.

Without options

         my $lxs = new Sys::Statistics::Linux;

Or with options

         my $lxs = Sys::Statistics::Linux->new(CpuStats => 1);

Would do nothing

         my $lxs = Sys::Statistics::Linux->new(CpuStats => 0);

It's possible to call C<new()> with a hash reference of options.

         my %options = (
            CpuStats => 1,
            MemStats => 1
         );

         my $lxs = Sys::Statistics::Linux->new(\%options);

Take a look to C<set()> for more informations.

=head2 set()

Call C<set()> to activate or deactivate options. The following example would call C<new()> and C<init()>
of C<Sys::Statistics::Linux::CpuStats> and delete the object of C<Sys::Statistics::Linux::SysInfo>:

         $lxs->set(
            CpuStats  => -1, # activated, but paused, wouldn't delete the object
            Processes =>  0, # deactivate - would delete the statistics and destroy the object
            PgSwStats =>  1, # activate the statistic and calls C<new()> and C<init()> if necessary
            NetStats  =>  2, # activate the statistic and call C<new()> if necessary but not C<init()>
         );

It's possible to call C<set()> with a hash reference of options.

         my %options = (
            CpuStats => 2,
            MemStats => 2
         );

         $lxs->set(\%options);

=head2 get()

Call C<get()> to get the collected statistics. C<get()> returns the statistics as a hash reference.

         my $stats = $lxs->get;

=head2 init()

The call of C<init()> re-init all statistics that are necessary for deltas and if the option is higher than 0.

         $lxs->init;

=head2 search(), fproc()

Both methods provides a simple search engine to find special statistics. Both methods except a filter
as a hash reference as the first argument. If your data comes from extern - maybe from a client that
send his statistics to the server - you can set the statistics as the second argument. The second
argument have to be a hash reference as well.

The C<search()> method search for statistics and rebuilds the hash tree until that keys that matched
your filter and returns the hits as a hash reference.

        my $hits = $lxs->search({
           Processes => {
              cmd   => qr/\[su\]/,
              owner => qr/root/
           },
           CpuStats => {
              total  => 'gt:50',
              iowait => '>10'
           },
           DiskUsage => {
              '/dev/sda1' => {
                 usageper => 'gt:80'
              }
           }
        });

This would return the following matches:

    * processes with the command C<[su]>
    * processes with the owner C<root>
    * all cpu where the C<total> usage is grather than 50
    * all cpu where C<iowait> is grather than 10
    * only that disk where C<usageper> is higher than 80

If your statistics comes from extern, than you can set the full statistics as the second argument.

        my %stats = (
           CpuStats => {
              cpu => {
                 system => '51.00',
                 total  => '51.00',
                 idle   => '49.00',
                 nice   => '0.00',
                 user   => '0.00',
                 iowait => '0.00'
              }
           }
        );
               
        my %filter = (
           CpuStats => {
              total => 'gt:50'
           }
        );

        my $hits = $lxs->search(\%filter, \%stats);

The C<fproc()> method search for processes only and returns a array reference with all process IDs that
matched the filter. Example:

        my $pids = $lxs->fproc({ cmd => qr/init/, owner => 'eq:apache' });

You can set the statistics as second argument as well if your statistics comes from extern.

        my $pids = $lxs->fproc(\%filter, \%stats);

This would return the following process ids:

    * processes that matched the command C<init>
    * processes with the owner C<apache>

There are different match operators available:

    * gt or > - grather than
    * lt or < - less than
    * eq or = - is equal
    * ne or ! - is not equal

Notation examples:

    gt:50  or  >50
    lt:50  or  <50
    eq:50  or  =50
    ne:50  or  !50

Both argumnents have to be set as a hash reference.

=head2 settime()

Call C<settime()> to define a POSIX formatted time stamp, generated with localtime().

         $lxs->settime('%Y/%m/%d %H:%M:%S');

To get more informations about the formats take a look at C<strftime()> of POSIX.pm
or the manpage C<strftime(3)>.

=head2 gettime()

C<gettime()> returns a POSIX formatted time stamp, @foo in list and $bar in scalar context.
If the time format isn't set then the default format "%Y-%m-%d %H:%M:%S" will be set automatically.
You can also set a time format with C<gettime()>.

         my $date_time = $lxs->gettime;

Or

         my ($date, $time) = $lxs->gettime;

Or

         my ($date, $time) = $lxs->gettime('%Y/%m/%d %H:%M:%S');

=head1 EXAMPLES

A very simple perl script could looks like this:

         use warnings;
         use strict;
         use Sys::Statistics::Linux;

         my $lxs = Sys::Statistics::Linux->new( CpuStats => 1 );
         sleep(1);
         my $stats = $lxs->get;
         my $cpu   = $stats->{CpuStats}->{cpu};

         print "Statistics for CpuStats (all)\n";
         print "  user      $cpu->{user}\n";
         print "  nice      $cpu->{nice}\n";
         print "  system    $cpu->{system}\n";
         print "  idle      $cpu->{idle}\n";
         print "  ioWait    $cpu->{iowait}\n";
         print "  total     $cpu->{total}\n";

Example to collect network statistics with a nice output:

         use warnings;
         use strict;
         use Sys::Statistics::Linux;

         $| = 1;

         my $header  = 20;
         my $average = 1;
         my $columns = 8;
         my $options = { NetStats => 1 };

         my @order = qw(
            rxbyt rxpcks rxerrs rxdrop rxfifo rxframe rxcompr rxmulti
            txbyt txpcks txerrs txdrop txfifo txcolls txcarr txcompr
         );

         my $lxs = Sys::Statistics::Linux->new( $options );

         my $h = $header;

         while (1) {
            sleep($average);
            my $stats = $lxs->get;
            if ($h == $header) {
               printf "%${columns}s", $_ for ('iface', @order);
               print "\n";
            }
            foreach my $device (keys %{$stats->{NetStats}}) {
               my $dstat = $stats->{NetStats}->{$device};
               printf "%${columns}s", $device;
               printf "%${columns}s", $dstat->{$_} for @order;
               print "\n";
            }
            $h = $header if --$h == 0;
         }

Activate and deactivate statistics:

         use warnings;
         use strict;
         use Sys::Statistics::Linux;
         use Data::Dumper;

         my $lxs = new Sys::Statistics::Linux;

         # set the options
         $lxs->set(
            SysInfo  => 1,
            CpuStats => 1,
            MemStats => 1
         );

         # sleep to get useful statistics for CpuStats
         sleep(1);

         # $stats contains SysInfo, CpuStats and MemStats
         my $stats = $lxs->get;
         print Dumper($stats);

         # we deactivate CpuStats
         $lxs->set(SysStats => 0);

         # $stats contains CpuStats and MemStats
         sleep(1);
         $stats = $lxs->get;
         print Dumper($stats);

Set and get a time stamp:

         use warnings;
         use strict;
         use Sys::Statistics::Linux;

         my $lxs = new Sys::Statistics::Linux;
         $lxs->settime('%Y/%m/%d %H:%M:%S');
         print "$lxs->gettime\n";

If you're not sure you can use the the C<Data::Dumper> module to learn more about the hash structure:

         use warnings;
         use strict;
         use Sys::Statistics::Linux;
         use Data::Dumper;

         my $lxs = Sys::Statistics::Linux->new( CpuStats => 1 );
         sleep(1);
         my $stats = $lxs->get;

         print Dumper($stats);

Take a look into the the F<examples> directory of the distribution for some examples with a nice output. :-)

=head1 DEPENDENCIED

    Test::More
    Carp

=head1 EXPORTS

No exports.

=head1 TODOS

   * Dynamic loader for options/modules.
   * Maybe Sys::Statistics::Linux::Formatter to format statistics
     for inserts into a database or a nice output to files.
   * Are there any wishs from your side? Send me a mail!

=head1 REPORTING BUGS

Please report all bugs to <jschulz.cpan(at)bloonix.de>.

=head1 AUTHOR

Jonny Schulz <jschulz.cpan(at)bloonix.de>.

=head1 COPYRIGHT

Copyright (c) 2006, 2007 by Jonny Schulz. All rights reserved.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut

package Sys::Statistics::Linux;
our $VERSION = '0.09_14';

use strict;
use warnings;
use Carp qw(croak);
use POSIX qw(strftime);
use constant RXOPTION => qr/^[0-2\-1]$/;

sub new {
   my $class = shift;
   my $self = bless {
      opts  => {
         SysInfo   =>  0,
         CpuStats  =>  0,
         ProcStats =>  0,
         MemStats  =>  0,
         PgSwStats =>  0,
         NetStats  =>  0,
         SockStats =>  0,
         DiskStats =>  0,
         DiskUsage =>  0,
         LoadAVG   =>  0,
         FileStats =>  0,
         Processes =>  0,
      },
      init  => {},
      stats => {},
      obj   => {},
   }, $class; 
   $self->set(@_) if @_;
   return $self;
}

sub set {
   my $self  = shift;
   my $class = ref $self;
   my $sets  = $class->_struct(@_);
   my $opts  = $self->{opts};
   my $obj   = $self->{obj};
   my $stats = $self->{stats};

   foreach my $opt (keys %{$sets}) {

      # validate the options
      croak "$class: invalid option '$opt'"
         unless exists $opts->{$opt};
      croak "$class: invalid value for '$opt'"
         unless $sets->{$opt} =~ RXOPTION;

      $opts->{$opt} = $sets->{$opt};

      if ($opts->{$opt} > 0) {
         my $package = $class."::".$opt;

         # require mod if not loaded
         unless (defined &{$package.'::new'}) {
            my $require = $package;
            $require =~ s/::/\//g;
            require "$require.pm"
               or croak "$class: unable to load $require.pm";
         }

         # create a new object if the object doesn't exist
         $obj->{$opt} = $package->new()
            unless $obj->{$opt};

         # get initial statistics if init() is defined and the
         # option is set to 1
         $obj->{$opt}->init() 
            if $opts->{$opt} == 1
            && defined &{$package.'::init'};

      } elsif ($opts->{$opt} == 0) {
         delete $obj->{$opt}   if $obj->{$opt};
         delete $stats->{$opt} if $stats->{$opt};
      }
   }
}

sub init {
   my $self  = shift;
   my $class = ref $self;
   my $obj   = $self->{obj};
   my $opts  = $self->{opts};

   foreach my $opt (keys %{$opts}) {
      my $package = $class."::".$opt;
      $obj->{$opt}->init()
         if defined &{$package.'::init'}
         && $opts->{$opt} > 0;
   }
}

sub settime {
   my $self   = shift;
   my $format = $_[0] ? $_[0] : '%Y-%m-%d %H:%M:%S';
   $self->{timeformat} = $format;
}

sub gettime {
   my $self = shift;
   $self->settime(@_) unless $self->{timeformat};
   my $tm = strftime($self->{timeformat}, localtime);
   return wantarray ? split /\s+/, $tm : $tm;
}

sub get {
   my $self   = shift;
   my $opts   = $self->{opts};
   my $stats  = $self->{stats};
   my $obj    = $self->{obj};
   my $format = $self->{timeformat};

   foreach my $opt (keys %{$opts}) {
      $stats->{$opt} = $obj->{$opt}->get()
         if $opts->{$opt} > 0;
   }

   return $stats;
}

sub search {
   my $self   = shift;
   my $class  = ref($self);

   croak "$class: first argument have to be a hash ref"
     unless $_[0] && ref($_[0]) eq 'HASH';
   croak "$class: second argument have to be a hash ref"
     if $_[1] && ref($_[1]) ne 'HASH';

   my ($filter, $stats) = @_ == 2 ? @_ : (shift, $self->{stats});

   # $stats and $filter must be set
   return undef unless %{$stats} && %{$filter};

   my $opts   = $self->{opts};
   my %hits   = ();

   foreach my $opt (keys %{$filter}) {

      # this 2 croaks are the only croaks in this loop
      croak "$class: not a hash ref opt '$opt'"
         unless ref($filter->{$opt}) eq 'HASH';
      croak "$class: invalid option '$opt'"
         unless exists $opts->{$opt};

      # next if the object isn't loaded
      next unless exists $stats->{$opt};

      # some sub refs
      my $fref = $filter->{$opt};
      my $sref = $stats->{$opt};

      # we search for matches for each key that is defined
      # in %filter and rebuild the tree until that key that
      # matched the searched string

      foreach my $x (keys %{$fref}) {

         # if $fref->{$x} is a hash ref than the next key have to
         # match the statistic key. this is used for statistics
         # like NetStats or Processes that uses a hash key for the
         # device name or process id. NetStats example:
         #
         # if
         #
         #    $fref->{eth0}->{ttbyt}
         #
         # is defined as filter, than the key "eth0" have to match
         #
         #    $sref->{eth0}
         #
         # than we look if "ttbyt" matched the searched string

         if (ref($fref->{$x}) eq 'HASH') {

            # if the key $sref->{eth0} doesn't exists
            # than we continue with the next defined filter
            next unless exists $sref->{$x};

            while ( my ($name, $value) = each %{$fref->{$x}} ) {
               $hits{$opt}{$x}{$name} = $sref->{$x}->{$name}
                  if exists $sref->{$x}->{$name}
                  && $class->_diff($sref->{$x}->{$name}, $value);
            }

         # if ref($fref->{$x}) is return a regexp or 'nothing' than we
         # search for matches in all key-value pairs that matched

         } elsif (ref($fref->{$x}) =~ /^(Regexp|)$/) {
            foreach my $key (keys %{$sref}) {
               if (ref($sref->{$key}) eq 'HASH') {
                  $hits{$opt}{$key}{$x} = $sref->{$key}->{$x}
                     if exists $sref->{$key}->{$x}
                     && $class->_diff($sref->{$key}->{$x}, $fref->{$x});
               } else {
                  $hits{$opt}{$x} = $sref->{$x}
                     if exists $sref->{$x}
                     && $class->_diff($sref->{$x}, $fref->{$x});
                  last;
               }
            }
         }
      }
   }

   return %hits ? \%hits : undef;
}

sub fproc {
   my $self   = shift;
   my $class  = ref($self);

   croak "$class: first argument have to be a hash ref"
     unless $_[0] && ref($_[0]) eq 'HASH';
   croak "$class: second argument have to be a hash ref"
     if $_[1] && ref($_[1]) ne 'HASH';

   my ($filter, $stats) = @_ == 2 ? @_ : (shift, $self->{stats});

   return undef unless %{$stats->{Processes}} && %{$filter};

   my @hits = ();
   my $sref = $stats->{Processes};

   foreach my $pid (keys %{$sref}) {
      my $proc = $sref->{$pid};

      while ( my ($key, $value) = each %{$filter} ) {
         push @hits, $pid
            if exists $proc->{$key}
            && $class->_diff($proc->{$key}, $value);
      }
   }

   return @hits ? \@hits : undef;
}

#
# private stuff
#

sub _diff {
   my ($c, $x, $y) = @_;

   return 1
      if ( ref($y) eq 'Regexp'  &&  $x =~ $y )
      || ( $y =~ s/^(eq:|=)//   &&  $x eq $y )
      || ( $y =~ s/^(ne:|!)//   &&  $x ne $y )
      || ( $y =~ s/^(gt:|>)//   &&  $y =~ /^\d+$/  &&  $x > $y )
      || ( $y =~ s/^(lt:|<)//   &&  $y =~ /^\d+$/  &&  $x < $y );

   return undef;
}

sub _struct {
   my $class = shift;

   return
      ref($_[0])
         ? ref($_[0]) eq 'HASH'
            ? $_[0]
            : croak "not a hash ref"
         : @_ % 2
            ? croak "odd number of elements in hash"
            : {@_};
}

1;
